### ISSUE TITLE
address sanitizer does not detect buffer overflow in global variable

### ISSUE THREAD

**Issue by yamam** [google/sanitizers#1588]
The following example from https://github.com/google/sanitizers/wiki/AddressSanitizerExampleGlobalOutOfBounds will detect a buffer overflow.

```c
// RUN: clang -O -g -fsanitize=address %t && ./a.out
int global_array[100] = {-1};
int main(int argc, char **argv) {
  return global_array[argc + 100];  // BOOM
}
```

However, if I change the access position from 100 to 200, the buffer overflow is not detected. Is there any way to make it detectable in this case?

```c
// RUN: clang -O -g -fsanitize=address %t && ./a.out
int global_array[100] = {-1};
int main(int argc, char **argv) {
  return global_array[argc + 200];  // BOOM
}
```

Environment
```
$ uname -srvmpio
Linux 5.15.0-52-generic #58-Ubuntu SMP Thu Oct 13 08:03:55 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
$ clang --version                                                                                                                                                                                    [~] 10/26 7:25
Ubuntu clang version 14.0.0-1ubuntu1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by ramosian-glider** [comment]
ASan uses redzones to detect buffer overflows, so if you jump over the redzone you won't be able to detect the overflow.
The redzone size starts at 32 bytes and is calculated depending on the size of the array itself.
You can increase the array size to get a bigger redzone, but it still won't help you to detect arbitrary buffer overflows.

**Comment by yamam** [comment]
Thank you for your prompt reply.
Is there any way to increase the redzone size?

**Comment by Enna1** [comment]
> I checked and could not find any compile or runtime options that would increase the redzone size. If there is an option to increase the redzone size, it would be beneficial to increase the detectable range, although it would increase memory usage.

Try `-fsanitize=address -mllvm --asan-mapping-scale=7`.
Here is an example: https://godbolt.org/z/nP4YsxbEG
Unfortunately, it gives us SEGV on unknown address not global-buffer-overflow.

**Comment by ramosian-glider** [comment]
No there is no easy way to increase the redzone size.
Our experience shows this is impractical, because in most cases out-of-bounds accesses hit the memory immediately following the variable. On the other hand, for any chosen redzone size there still will be cases which it won't help catch.

**Comment by ramosian-glider** [comment]
> Try `-fsanitize=address -mllvm --asan-mapping-scale=7`. Here is an example: https://godbolt.org/z/nP4YsxbEG Unfortunately, it gives us SEGV on unknown address not global-buffer-overflow.

`--asan-mapping-scale` is not officially supported. I don't think ASan runtime accounts on it being different from 3.

**Comment by Enna1** [comment]
> > Try `-fsanitize=address -mllvm --asan-mapping-scale=7`. Here is an example: https://godbolt.org/z/nP4YsxbEG Unfortunately, it gives us SEGV on unknown address not global-buffer-overflow.
> 
> `--asan-mapping-scale` is not officially supported. I don't think ASan runtime accounts on it being different from 3.

Yes, I have seen some places in ASan runtime with hardcoded mapping-scale to 3.

**Comment by yamam** [comment]
I tried it with --asan-mapping-scale=7.
Even accesses within the range resulted in SEGV.
https://godbolt.org/z/bG183Ynjn

I was unable to detect a buffer overflow bug that greatly exceeded the range in ASan, so I looked for a way to detect it, but unfortunately it doesn't seem to be possible. Thank you very much.
